### PR TITLE
Support openaccess product Id, infer isAccessibleForFree from product Id if not explicitly set

### PR DESCRIPTION
**Pull request by justinchou-google**
TLDR: for June launch, require contribution publishers to have in snippet: `isAccessibleForFree: true`; else, infer `isAccessibleForFree` from product Id. Background: b/280335205

Currently, `isAccessibleForFree` is utilized in (at least) two flows:

[b/277725085] `AutoPromptManager` - `isAccessibleForFree` sets whether prompt `isClosable` or not. Adds support for `isAccessibleForFree` to not be set, and `isClosable` is inferred from whether publication is contribution or subscription (based on article response). Renames `isAccessibleForFree` in autoPrompt flow to `isClosable`.

[b/277725876] `pageConfig` - `isAccessibleForFree` is an input to whether page `isLocked`, which is read downstream in `autoPromptManager`, `entitlements-manager`, and `gaa-metering`. Still requires `isAccessibleForFree` to be set for contribution publishers. If `isAccessibleForFree` is not set, assume the revenue model is not contribution, and we will infer from the productId: `openacess` productId will set `isAccessibleForFree: true`, else set `isAccessibleForFree: false`.

### CODE REVIEW

**Review by ChrisAntaki — APPROVED** [review]

File: src/runtime/basic-runtime-test.js [hunk 1]
```diff
@@ -261,6 +263,116 @@ describes.realWin('BasicRuntime', (env) => {
...
+        isAccessibleFromProductId: true,
+      },
+      {
+        isAccessibleForFree: undefined,
+        isPartOfProductId: 'publication:notopen',
+        isAccessibleFromProductId: false,
+      },
+    ].forEach(
+      ({isAccessibleForFree, isPartOfProductId, isAccessibleFromProductId}) => {
+        it(`When isAccessibleForFree=${isAccessibleForFree}, isPartOfProductId=${isPartOfProductId}, expect writePageConfig to be called with isAccessibleForFree=${isAccessibleFromProductId}`, async () => {
```
Comment (fikus) [inline src/runtime/basic-runtime-test.js:300]: nit: Can you write these in the typical `it` style? eg, it('writes page config with isAccesibleForFree=x when isAccessibleForFree=y and isPartOfProductId=z')

Comment (ChrisAntaki) [reply src/runtime/basic-runtime-test.js:300]: +1

Comment (justinchou-google) [reply src/runtime/basic-runtime-test.js]: Done!

File: src/runtime/basic-runtime.ts [hunk 2]
```diff
@@ -207,27 +220,28 @@ export class BasicRuntime implements BasicSubscriptions {
...
-        this.pageConfigWriter_ = null;
-        this.configured_(true);
-      });
+    const isOpenAccess = isOpenAccessProductId(isPartOfProductId);
+    let isClosable = isAccessibleForFree;
+    // Only overwrite isClosable if product is openaccess, else leave undefined.
+    if (isOpenAccess) {
+      isClosable ??= true;
+    }
+    isAccessibleForFree ??= isOpenAccess;
```
Comment (oyj9109) [inline src/runtime/basic-runtime.ts:230]: Should this be left undefined if product is not openaccess too?

**Review by oyj9109 — COMMENTED** [review]
Let's wait for the snippet discussion with John next week.

File: src/runtime/basic-runtime.ts (hunk 2 above)
Comment (justinchou-google) [reply src/runtime/basic-runtime.ts]: resolving as discussed in b/280335205.

File: src/runtime/basic-runtime.ts [hunk 3]
```diff
@@ -211,27 +224,29 @@ export class BasicRuntime implements BasicSubscriptions {
...
+      isPartOfType,
+      isPartOfProductId,
+    });
 
     this.clientOptions_ = Object.assign({}, clientOptions, {
       forceLangInIframes: true,
     });
+
+    let isClosable = isAccessibleForFree;
+    // Only overwrite isClosable if product is openaccess, else leave undefined.
```
Comment (ChrisAntaki) [inline src/runtime/basic-runtime.ts:242]:
```suggestion
    // Only default isClosable to true if product is openaccess, else leave undefined.
```

File: src/runtime/basic-runtime.ts [hunk 4]
```diff
@@ -189,6 +189,19 @@ export class BasicRuntime implements BasicSubscriptions {
...
   }
 
+  private writePageConfig_(markupValues: {
+    type: string | Array<string>;
+    isAccessibleForFree: boolean;
+    isPartOfType: string | Array<string>;
+    isPartOfProductId: string;
+  }): void {
+    this.pageConfigWriter_ = new PageConfigWriter(this.doc_);
+    this.pageConfigWriter_.writeConfigWhenReady(markupValues).then(() => {
```
Comment (ChrisAntaki) [inline src/runtime/basic-runtime.ts:199]: Could we flatten this with async/await?

Comment (ChrisAntaki) [inline src/runtime/basic-runtime.ts:199]: Could we flatten this with async/await?
```suggestion
    await this.pageConfigWriter_.writeConfigWhenReady(markupValues);
```

**Review by ChrisAntaki — APPROVED** [review]
Looks good! Just added 2 comments

**Review by oyj9109 — APPROVED** [review]

**Comment by oyj9109** [comment]
Looks great. Thanks for the investigation and detailed summary.
